### DISCUSSION TITLE
Do I have to exactly have the both words "airflow" and "DAG" in the file to be parsed and registered as DAG?

### DISCUSSION THREAD

**Discussion by PDHBE** [apache/airflow#24458] [Q&A]
### Apache Airflow version

2.2.4

### What happened

The DAG I created is not registered.

I'm using DAG Class and Operator by extending as follows:
`class DefaultDAG(DAG):`
`class BatchJobOperator(BaseOperator):`

So I'm not using :
`from airflow import DAG`
`from airflow.operators.python import PythonOperator`
`DAG(...)`

I'm using:
`from batch.dags.default_dag import DefaultDAG`
`from common.operator.batch_job_operator import BatchJobOperator`
`DefaultDAG(...)`

Therefore, my DAG file exactly contain the string "DAG" but not "airflow"

And I found the following in the official document:
`Does the file containing your DAG contain the string “airflow” and “DAG” somewhere in the contents? When searching the DAG directory, Airflow ignores files not containing “airflow” and “DAG” in order to prevent the DagBag parsing from importing all python files collocated with user’s DAGs.`

Is it correct that it is not recognized as DAG because of this reason?

And I added only one line of `from airflow import DAG` for testing, and the DAG is recognized successfully.

And I Found "guesses whether a Python file contains an Airflow DAG definition" Logic:
```
def might_contain_dag(file_path: str, safe_mode: bool, zip_file: Optional[zipfile.ZipFile] = None):
    """
    Heuristic that guesses whether a Python file contains an Airflow DAG definition.

    :param file_path: Path to the file to be checked.
    :param safe_mode: Is safe mode active?. If no, this function always returns True.
    :param zip_file: if passed, checks the archive. Otherwise, check local filesystem.
    :return: True, if file might contain DAGs.
    """
    if not safe_mode:
        return True
    if zip_file:
        with zip_file.open(file_path) as current_file:
            content = current_file.read()
    else:
        if zipfile.is_zipfile(file_path):
            return True
        with open(file_path, 'rb') as dag_file:
            content = dag_file.read()
    content = content.lower()
    return all(s in content for s in (b'dag', b'airflow'))

```

### What you think should happen instead

If the reason is that both strings "airflow" and "DAG" are not included exactly,
I think it is limited to be recognized as DAG only if both strings "airflow" and "DAG" are included exactly.

Also, unlike the "DAG Import Errors" that can be clearly seen through the web UI,
The cause of DAG not being recognized is not clearly indicated, so it is difficult to figure out what is the problem.

@seunggabi

### Operating System

Debian GNU/Linux 10 (buster)

### Deployment

Other Docker-based deployment

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by boring-cyborg** [comment]
Thanks for opening your first issue here! Be sure to follow the issue template!

**Comment by potiuk** [comment]
Yes. You have to add DAG. You can do it in comments.

**Reply by ashb** [reply]
Correction, you have to do this _by default_ but it is configurable.

There is a setting ` [core]` called `dag_discovery_safe_mode` -- set that to False and it will load/parse/execute every python file it finds under the dags folder. (hence why it's "not as safe")

**Reply by potiuk** [reply]
TIL. Thanks @ashb !

**Comment by potiuk** [comment]
If you have an idea how to improve it so that it is more visible - Pull Request is most welcome.

**Reply by seunggabi** [reply]
@potiuk 
hmm .. I have a good idea.
file name or directory name has `dag`?

1. many developer use `dag` keyword.
2. file count is smaller than text .. haha..;

```
dags/asdf.py
asdf/asdf_dag.py
```

if you okay, I make a code.

**Reply by potiuk** [reply]
Let's see what others say. I am not totally against, though it will require backwards compatibilty and be disabled by default to be considered at all.

**Reply by seunggabi** [reply]
okay as soon as possible make pr!
thx your opinion!

**Reply by tomasfarias** [reply]
I'd like to raise a potential problem with DAG generation tools. Take, for example, the basic example presented by [dag-factory](https://github.com/ajbosco/dag-factory):

```python
from airflow import DAG
import dagfactory

dag_factory = dagfactory.DagFactory("/path/to/dags/config_file.yml")

dag_factory.clean_dags(globals())
dag_factory.generate_dags(globals())
```

The first `import` is not used for anything except satisfying `might_contain_dag`. Any folks running `flake8` or similar tools in their CI/CD may be inclined to remove the line to shut off the F401 error, and cause their DAGs to disappear.

Because of this, I think @seunggabi's ideas are valuable: all of our dags are stored in a `dags` folder and/or have a `_dag` suffix in their file name.

Perhaps a `if 'dag' in file_path: return True` could be added to `might_contain_dag` in a similar fashion to how it now checks whether the file is a zip file?

Thanks to all for this discussion!

**Reply by potiuk** [reply]
You can add one line:

```
# DAG
```

The DAG might be also in comments. No problem with that. If you generate it from a tool, it might also be:

```
# This DAG file has been generated via my_awesome_dag_generation tool on 2022-12-31 23:59:59 by tomasfarias
```

Which is also great for traceability, readability etc.

No problem with flake8 or any other tools.

**Comment by emmmwhammm** [comment]
I wasted hours on this - genuinely really annoying

**Reply by potiuk** [reply]
Then you will be likely heppy to know that in Airflow 3 we are introducing DAG bundle concept, where the whole idea on how you determine what and where is parsed will be much more localized and improved.

I suggest you subscribe to devlist and try it when Airflow 3 is out and help us to test it  and provide feedback.

This is the best way you can help to improve this for you and others. Way better than venting frustration - doing something is so much better.

Can we count on your help?
